PSKoans/Koans/IntroTraining/AboutGetMember.koans.ps1

```diff
@@ -0,0 +1,167 @@
+
+
+   TypeName: System.ServiceProcess.ServiceController
+
+    Notice the "TypeName" is System.ServiceProcess.ServiceController. This tells us that
+    the Get-Service cmdlet returns objects of a ServiceController type. They reside in
+    the System.ServiceProcess namespace but you don't need to know more about this unless
+    you dive into .NET development.
+
+    In these Koans we'll go over the Get-Member cmdlet,
+    expanding upon our knowledge of object-orientated program, in order to find properties
+    and methods of an object. The next kew Koans will relate to what these "TypeName" items are.
+
+#>
+
+$RelativePath = ".\AnswerFunctions\IntroTraining\AboutGetMemberAnswers\AboutGetMemberAnswers.psd1"
+Import-Module $RelativePath
+
+
+Describe 'Get-Member' {
+<#
+    The Get-Member cmdlet is super useful.
+
+    If you remember back in "AboutDotNet" we discovered that PowerShell is object-orientated;
+    everything is an object with properties (things the object has) and methods (things
+    the object can do). 
+
+    When we use Get-Member on any object (everything is an object in PowerShell!) it'll return to you
+    the datatype (TypeName) and a list of properties and methods the object has.
+
+    That's pretty neat. Using this not only can we figure out what type of object a 
+    thing is, but we can also see all of the information available to us from it and
+    any actions we can do with it.
+
+    So we can get properties and methods via Get-Member, and we know what these things are,
+    but how do we access them? .NET has a concept called dot notation, which boils down to:
+
+    Object.property
+    Object.method()
+
+    We'll cover why we have to use brackets with methods later.
+
+    For now, use Get-Member - plus dot notation - to retrieve a few methods and properties.
+#>
+
+    It 'Propeties' {
+
+        # Retrieve properties returned by various cmdlets.
+        # See below for an example:
+        <# EXAMPLE
+        # Specify the cmdlet here.
+        $cmdlet = Get-Service
+        # And the name of the property you've found
+        # E.G. "__" becomes "CanPauseAndContinue"
+        $PropertyName = "CanPauseAndContinue"
+        VerifyMember -MemberType Property -MemberName $PropertyName -Object $cmdlet | Should -BeTrue `
+        -Because "$PropertyName property doesn't exist in given cmdlet $cmdlet"
+        #>
+        
+
+        # Below, specify three cmdlets and some of their properites you've discovered.
+
+        ##### 01 #####
+        # Specify the cmdlet here.
+        # e.g. "__" becomes Get-Service
+        $cmdlet = "__"
+        # And the name of the property you've found
+        # E.G. "__" becomes "CanPauseAndContinue"
+        $PropertyName = "__"
+        VerifyMember -MemberType Property -MemberName $PropertyName -Object $cmdlet | Should -BeTrue `
+        -Because "$PropertyName property doesn't exist in given cmdlet $cmdlet"
```

Comment:
I think we can simplify all of this, and remove the need to have comments in a lot of it. Most of the comments in this section are repeated, so I think we can at least bring it down to one set of comments, if nothing else.

Looking at this, I think we can frame these a little more effectively by:

1. Having only the most necessary comments in this section (the prior comments are great, just a few too many in and around where the user is expected to be working, I think!)
2. Avoiding the abstraction. I tend to try to stick to just commands users can work with directly in PowerShell; it encourages exploration, and it shows them how things work as we go along, rather than showing them a command that is more or less a black box unless they go hunting for it. Don't be afraid to put together the verification for the answer in the koan itself. 🙂 
3. Explicitly checking the three `$cmdlet` inputs aren't the same (just name them slightly differently and then check them all at the end of the `It` and verify they're all different.

Oh, and as I mentioned above, any examples here should probably stick to commands that are available on most or all platforms; Get-Service is Windows-specific.

Reply:
1. Taken out comment block for ```It``` sections and made the overall Koan one more informative. ```It``` blocks have one comment block with an example, then no more comments.

All ```Get-Service``` cmdlet examples replaced with ```Get-Process``` examples.

Currently working on number 3 and to an extend number 2. Maybe I haven't woken up yet but for whatever reason this isn't working as a Koan but does as in native PowerShell:

## Koan
```PowerShell
        $cmdlet1 = 'Get-Service'
        $PropertyName = "Status"
        $Members =  Get-Service | Get-Member -MemberType Property -Name $PropertyName
        $Members.count -ne 0 | Should -BeTrue `
        -Because "$PropertyName property doesn't exist in given cmdlet $cmdlet1"

        $cmdlet2 = 'Get-Process'
        $PropertyName = "Threads"
        $Members =  Get-Process | Get-Member -MemberType Property -Name $PropertyName
        $Members.count -ne 0 | Should -BeTrue `
        -Because "$PropertyName property doesn't exist in given cmdlet $cmdlet2"

        $cmdlet3 = 'Get-ScheduledTask'
        $PropertyName = "Version"
        $Members =  Get-ScheduledTask | Get-Member -MemberType Property -Name $PropertyName
        $Members.count -ne 0 | Should -BeTrue `
        -Because "$PropertyName property doesn't exist in given cmdlet $cmdlet3"

        $Cmdlets = $cmdlet1, $cmdlet2, $cmdlet3
        $UniqueCheck = $Cmdlets | Get-Unique
        $UniqueCheck -eq 3 | Should -BeTrue `
        -Because "three unique cmdlets should be supplied"
```
## Result
```PowerShell
PS C:\Users\Administrator\Documents\GitHub\Dev\PSKoans> Measure-Karma -Topic AboutGetMember
    Welcome, seeker of enlightenment.
    Please wait a moment while we examine your karma...

Describing 'Get-Member' has damaged your karma.

    You have not yet reached enlightenment.

    The answers you seek...

Expected $true, because three unique cmdlets should be supplied, but got $null.

    Please meditate on the following code:

× It Propeties
at <ScriptBlock>, C:\Users\Administrator\PSKoans\IntroTraining\AboutGetMember.koans.ps1: line 92
92:         $UniqueCheck -eq 3 | Should -BeTrue `

    ▌ Knowledge and ignorance are interdependent;
    ▌ Delusion and enlightenment condition each other.

    You examine the path beneath your feet...



 [Total]: [―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――] 0/2

Run 'Show-Karma -Meditate' to begin your meditations.
```

## Native PowerShell
```PowerShell
function CheckMember() {
param(
[parameter(mandatory=$true)][String]$CmdletName,
[parameter(mandatory=$true)][Int]$MemberCount
)
    if ($MemberCount -ne 0) {
        Write-Host $("$CmdletName member found")
    } else {
        Write-Host $("$CmdletName member not found")
    }
}

$Cmdlet1 = 'Get-Service'
$PropertyName = "Status"
$Members =  Get-Service | Get-Member -MemberType Property -Name $PropertyName
CheckMember -CmdletName $Cmdlet1 -MemberCount $members.count

$cmdlet2 = 'Get-Process'
$PropertyName = "Threads"
$Members =  Get-Process | Get-Member -MemberType Property -Name $PropertyName
CheckMember -CmdletName $Cmdlet2 -MemberCount $members.count

$Cmdlet3 = 'Get-ScheduledTask'
$PropertyName = "Version"
$Members =  Get-ScheduledTask | Get-Member -MemberType Property -Name $PropertyName
CheckMember -CmdletName $Cmdlet3 -MemberCount $members.count

$Cmdlets = $cmdlet1, $cmdlet2, $cmdlet3
$UniqueCheck = ($Cmdlets | Get-Unique).count -eq 3
Write-Host $("Unique check result is " + $UniqueCheck)
```

```PowerShell
Get-Service member found
Get-Process member found
Get-ScheduledTask member found
Unique check result is True
```

Reply:
`$UniqueCheck -eq 3 | Should -BeTrue` Looks to be the problem in the koan version; your other version is checking the count, which is correct. 🙂 

These all look great! The only thing I'd say is definitely avoid using backticks to continue lines. If the line is too long with the `-Because` added, just add a line break after the pipe and indent the next line like so:
```powershell
$Members.Count -ne 0 |
    Should -BeTrue -Because "$PropertyName property doesn't exist in given cmdlet $cmdlet1"
```

Reply:
Ah I knew it'd be a silly mistake I'd made somewhere.

Changed it all but now I'm getting a very different error. Testing what happens when you complete all ```It``` blocks I get this:

```PowerShell
PS C:\Users\Administrator\PSKoans> Show-Karma -Topic AboutGetMember
    Welcome, seeker of enlightenment.
    Please wait a moment while we examine your karma...

Show-MeditationPrompt : Cannot validate argument on parameter 'DescribeName'. The argument is null or empty. Provide an 
argument that is not null or empty, and then try the command again.
At C:\Users\Administrator\Documents\GitHub\Dev\PSKoans\PSKoans\Public\Show-Karma.ps1:91 char:35
+             Show-MeditationPrompt @Params
+                                   ~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Show-MeditationPrompt], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationError,Show-MeditationPrompt
```

And I've tried:
```PowerShell
Describe 'Get-Member' {
```
```PowerShell
Describe "Get-Member" {
```
Thoughts? One thing to note as well is that since the ```-Reset``` param for ```Measure-Karma``` has been removed I've really struggled with what combination of cmdlets to use in order to effectively reset my environment to test changes. Is it worth adding that to the contribution page to assist with development?

For my own sake, for example, after making a change I'd normally do ```Measure-Karma -Reset``` and then ```Measure-Karma -Meditate``` in order to load up a fresh copy and test. Since then I gather the way is to do ```Reset-PSKoan``` and then ```Measure-Karma -Meditate``` but this doesn't seem to work for me sometimes? 

For example, I'm trying to test the above error with ```Describe "Get Member```. My AboutGetMembers shows:

```PowerShell
Describe "Get Member" {
```

I run the below to get a fresh user copy:
```PowerShell
PS C:\Users\Administrator\Documents\GitHub\Dev\PSKoans> Reset-PSKoan
Confirm
Are you sure you want to perform this action?
Performing the operation "Reset-PSKoan" on target "Resetting "compares collections of objects" in AboutCompareObject".
[Y] Yes [A] Yes to All [N] No [L] No to All [S] Suspend [?] Help (default is "Yes"): a
PS C:\Users\Administrator\Documents\GitHub\Dev\PSKoans> Measure-Karma -Meditate
```
But this copy still shows:
```PowerShell
Describe "Get-Member" {
```

So then I try combining it with ```Update-PSKoan```:

```PowerShell
PS C:\Users\Administrator\Documents\GitHub\Dev\PSKoans> Update-PSKoan
PS C:\Users\Administrator\Documents\GitHub\Dev\PSKoans> Reset-PSKoan
Confirm
Are you sure you want to perform this action?
Performing the operation "Reset-PSKoan" on target "Resetting "compares collections of objects" in AboutCompareObject".
[Y] Yes [A] Yes to All [N] No [L] No to All [S] Suspend [?] Help (default is "Yes"): a
PS C:\Users\Administrator\Documents\GitHub\Dev\PSKoans> Measure-Karma -Meditate
```

OR

```PowerShell
PS C:\Users\Administrator\Documents\GitHub\Dev\PSKoans> Reset-PSKoan
Confirm
Are you sure you want to perform this action?
Performing the operation "Reset-PSKoan" on target "Resetting "compares collections of objects" in AboutCompareObject".
[Y] Yes [A] Yes to All [N] No [L] No to All [S] Suspend [?] Help (default is "Yes"): a
PS C:\Users\Administrator\Documents\GitHub\Dev\PSKoans> Update-PSKoan
PS C:\Users\Administrator\Documents\GitHub\Dev\PSKoans> Measure-Karma -Meditate
```

But still get the same result; AboutGetMember.koans.ps1 not being updated for the user.

Reply:
🤔 interesting. The error from Show-MeditationPrompt might be something I need to fix... hm. 

I'll have to examine what's going on when a topic is specified and completed... might be a parameter set issue. Can you file an issue on that? I'll check it out today.

/cc @indented-automation for issues with the Update/Reset functions -- any ideas, man? 🙂

Reply:
Awesome! And I found the issue with Show-Karma as well; it isn't handling completed koan topics properly.

Reply:
@Sudoblark expect those to be fixed in PRs shortly; I've already opened one for the Show-Karma / Measure-Karma issue. I'll merge it after I write a test to prevent regression. 🙂

Reply:
I don't mean to be such a pain I promise!

Reply:
Haha, don't worry about it. I'd much rather find the bugs here than after I push a release :joy:

Reply:
Merged #250 to resolve the issue with not handling completed topics properly; you can rebase or merge the master branch in to get that sorted if you like.

And now I'll get on with reviewing this 😊

Reply:
```PowerShell
PS C:\Users\Administrator\PSKoans> Show-Karma -Topic AboutGetMember
    Welcome, seeker of enlightenment.
    Please wait a moment while we examine your karma...

    Congratulations! You have taken the first steps towards enlightenment.

    You have completed: AboutGetMember

    You cast your gaze back upon the path that you have walked...

 [■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■] 2/2
PS C:\Users\Administrator\PSKoans>
```

Yay!